src/test/crypto_tests.cpp

```diff
@@ -184,6 +184,33 @@ static void TestChaCha20(const std::string &hex_message, const std::string &hexk
     }
 }
 
+static void TestChaCha20RFC8439(const std::string& hex_key, ChaCha20::Nonce96 nonce, uint32_t counter, const std::string& hex_expected_keystream, const std::string& hex_input, const std::string& hex_expected_output)
+{
+    auto key = ParseHex(hex_key);
+
+    if (!hex_expected_keystream.empty()) {
+        ChaCha20 c20(key.data());
+        c20.SetRFC8439(nonce, counter);
+        std::vector<unsigned char> keystream;
+        auto expected_keystream = ParseHex(hex_expected_keystream);
+        keystream.resize(expected_keystream.size());
+        c20.Keystream(keystream.data(), expected_keystream.size());
+        BOOST_CHECK_EQUAL(HexStr(keystream), hex_expected_keystream);
+    }
+
+    if (!hex_input.empty()) {
+        assert(hex_input.size() == hex_expected_output.size());
```

Comment:
tiny nit (probably only worth changing if you have to retouch): could check this assert unconditionally, to not allow test cases where `hex_input` is empty but `hex_expected_output` is not